torch/utils/data/dataloader.py

```diff
@@ -792,6 +890,223 @@ def _next_data(self):
+            _StateKeys.FETCHER_STATE: fetcher_state,
+            _StateKeys.DATASET_STATE: dataset_state,
+            _StateKeys.ITERATOR_FINISHED: self._finished,
+        }
+        return state_dict
+
+    def load_state_dict(self, state_dict):
+        """Load state from a checkpoint."""
+        assert (
+            self._NUM_YIELDED in state_dict
+        ), f"State doesn't contain key '{self._NUM_YIELDED}' expected for single process dataloader"
+
+        # Load sampler state
+        self._load_sampler_state(state_dict)
+
+        # Load common state
+        self._load_common_state(state_dict)
+
+        # Load dataset state (if applicable)
+        self._load_dataset_state(state_dict)
+
+        # Create fetcher
+        self._create_fetcher()
+
+        # Load dataset-kind-specific state
+        if self._dataset_kind == _DatasetKind.Iterable:
+            self._load_iterable_dataset_state(state_dict)
+        # Map datasets don't need additional loading
+
+        # Load finished state
+        self._finished = state_dict[_StateKeys.ITERATOR_FINISHED]
+
+    def _load_sampler_state(self, state_dict):
+        """Restore sampler/sampler_iter state."""
+        self._sampler_iter_yielded = state_dict[_StateKeys.SAMPLER_ITER_YIELDED]
+
+        if isinstance(self._index_sampler, Stateful):
+            # Restore sampler and create fresh iterator from it
+            _try_load_state_dict(
+                self._index_sampler, state_dict[_StateKeys.INDEX_SAMPLER_STATE]
+            )
+            self._sampler_iter = iter(self._index_sampler)
+        elif isinstance(self._sampler_iter, Stateful):
+            # Only restore iterator state
+            self._sampler_iter = iter(self._index_sampler)
+            if state_dict[_StateKeys.SAMPLER_ITER_STATE] is not None:
+                _try_load_state_dict(
+                    self._sampler_iter, state_dict[_StateKeys.SAMPLER_ITER_STATE]
+                )
+        else:
+            if not isinstance(self._index_sampler, _InfiniteConstantSampler):
+                # Fallback to fastforward
+                self._sampler_iter = itertools.islice(
+                    self._index_sampler, self._sampler_iter_yielded, None
+                )
+
+    def _load_common_state(self, state_dict):
+        """Load state common to all dataset kinds."""
+        self._num_yielded = state_dict[self._NUM_YIELDED]
+        self._IterableDataset_len_called = state_dict[
+            _StateKeys.ITERABLEDATASET_LEN_CALLED
+        ]
+        self._shared_seed = state_dict[_StateKeys.SHARED_SEED]
+
+    def _load_dataset_state(self, state_dict):
+        """Load dataset state if it's stateful."""
+        if (
+            _StateKeys.DATASET_STATE in state_dict
+            and state_dict[_StateKeys.DATASET_STATE] is not None
+            and isinstance(self._dataset, Stateful)
+        ):
+            _try_load_state_dict(self._dataset, state_dict[_StateKeys.DATASET_STATE])
```

Comment:
If `_try_load_state_dict` is already doing the None-check then can simplify:
```suggestion
        _try_load_state_dict(self._dataset, state_dict.get(_StateKeys.DATASET_STATE))
```